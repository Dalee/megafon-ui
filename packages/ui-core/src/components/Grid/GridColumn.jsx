import * as React from 'react';
import * as PropTypes from 'prop-types';
import './GridColumn.less';
import cnCreate from 'utils/cn';

const cn = cnCreate('mfui-grid-column');
class GridColumn extends React.Component {
    static propTypes = {
        wide: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
        desktop: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
        tablet: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
        mobile: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
        align: PropTypes.oneOf(['right', 'left', 'center']),
        grow: PropTypes.bool,
        flex: PropTypes.boolean,
        all: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
        children: PropTypes.oneOfType([
            PropTypes.arrayOf(PropTypes.element),
            PropTypes.element,
            PropTypes.string,
        ]),
    };

    static defaultProps = {
        all: 12,
        flex: false,
        grow: false,
    };

    render() {
        const { all, wide, desktop, tablet, className, flex, grow, align, children, mobile } = this.props;

        return (
            <div
                className={cn(
                    '',
                    {
                        flex,
                        grow,
                        align,
                        all,
                        wide,
                        desktop,
                        tablet,
                        mobile,
                    },
                    className
                )}
            >
                {children}
            </div>
        );
    }
}

export default GridColumn;
