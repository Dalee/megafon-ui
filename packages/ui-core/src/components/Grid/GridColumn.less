@import '~styles/base.less';

.grid-column {
    &_grow {
        flex-grow: 1;
    }
    &_align_right {
        margin-left: auto;
    }
    &_align_left {
        margin-right: auto;
    }
    &_align_center {
        margin-right: auto;
        margin-left: auto;
    }

    &_flex {
        display: flex;
        flex-shrink: 0;
        flex-grow: 0;
    }

    .generateColumns(all);

    @media @desktopMU {
        .generateColumns(wide);
    }
    @media @desktopS {
        .generateColumns(desktop);
    }
    @media @mobileB {
        .generateColumns(tablet);
    }
    @media @mobileSM {
        .generateColumns(mobile);
    }
}

.generateColumns(@resolution, @columnCount: 12, @i: 1) when (@i <= @columnCount) {
    &_@{resolution}_@{i} {
        width: (100% / @columnCount) * @i;
    }
    .generateColumns(@resolution, @columnCount, (@i + 1));
}
