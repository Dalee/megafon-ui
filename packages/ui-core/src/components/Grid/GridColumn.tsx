import * as React from 'react';
import * as PropTypes from 'prop-types';
import './GridColumn.less';
import cnCreate from 'utils/cn';

export interface IGridColumn {
    /** Size of columns on wide screens */
    wide?: '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9' | '10' | '11' | '12';
    /** Size of columns on desktop */
    desktop?: '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9' | '10' | '11' | '12';
    /** Size of columns on tablet */
    tablet?: '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9' | '10' | '11' | '12';
    /** Size of columns on mobile */
    mobile?: '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9' | '10' | '11' | '12';
    /** Size of columns on all screens */
    all?: '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9' | '10' | '11' | '12';
    /** Align columns */
    align?: 'right' | 'left' | 'center';
    /** Column flex grow */
    grow?: boolean;
    /** Column as flex container */
    flex?: boolean;
    /** Custom class name */
    className?: string;
    children: JSX.Element[] | Element[] | JSX.Element | string | Element;
}

const cn = cnCreate('mfui-grid-column');
class GridColumn extends React.Component<IGridColumn, {}> {
    static propTypes = {
        wide: PropTypes.oneOf(['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12']),
        desktop: PropTypes.oneOfType(['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12']),
        tablet: PropTypes.oneOfType(['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12']),
        mobile: PropTypes.oneOfType(['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12']),
        all: PropTypes.oneOfType(['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12']),
        align: PropTypes.oneOf(['right', 'left', 'center']),
        grow: PropTypes.bool,
        flex: PropTypes.bool,
        className: PropTypes.string,
        children: PropTypes.oneOfType([
            PropTypes.arrayOf(PropTypes.element),
            PropTypes.element,
            PropTypes.string,
        ]),
    };

    static defaultProps = {
        all: 12,
        flex: false,
        grow: false,
    };

    render() {
        const { all, wide, desktop, tablet, className, flex, grow, align, children, mobile } = this.props;

        return (
            <div
                className={cn(
                    '',
                    {
                        flex,
                        grow,
                        align,
                        all,
                        wide,
                        desktop,
                        tablet,
                        mobile,
                    },
                    className
                )}
            >
                {children}
            </div>
        );
    }
}

export default GridColumn;
