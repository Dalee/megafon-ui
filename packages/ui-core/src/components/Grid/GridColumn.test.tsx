import * as React from 'react';
import { shallow } from 'enzyme';
import GridColumn from './GridColumn';

describe('<GridColumn />', () => {
    it('it renders GridColumn without props', () => {
        const wrapper = shallow(<GridColumn>item</GridColumn>);
        expect(wrapper).toMatchSnapshot();
    });

    it('it renders GridColumn with props', () => {
        const wrapper = shallow(
            <GridColumn
                wide="8"
                desktop="4"
                tablet="3"
                mobile="1"
                all="10"
                align="right"
                grow
                flex
            >
                item
            </GridColumn>
        );
        expect(wrapper).toMatchSnapshot();
    });
});
