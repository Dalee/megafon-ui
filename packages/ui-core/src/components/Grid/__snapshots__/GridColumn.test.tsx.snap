// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GridColumn /> it renders GridColumn with props 1`] = `
<div
  className="mfui-grid-column mfui-grid-column_flex mfui-grid-column_grow mfui-grid-column_align_right mfui-grid-column_all_10 mfui-grid-column_wide_8 mfui-grid-column_desktop_4 mfui-grid-column_tablet_3 mfui-grid-column_mobile_1"
>
  item
</div>
`;

exports[`<GridColumn /> it renders GridColumn without props 1`] = `
<div
  className="mfui-grid-column mfui-grid-column_all_12"
>
  item
</div>
`;
